### PR TITLE
Add an optional toolip to SortableColumn headers.

### PR DESCRIPTION
**Pull request by nickshanks**
# Feature

## Summary
@TomGoodIdea said [this was a pre-requisite](https://github.com/endless-sky/endless-sky/pull/10250#issuecomment-2182656997) for PR #10250 
It adds a tooltip to the column headers on the fleet table of the player info panel.
I added just one tooltip, for what I think is the least intuitive column currently, crew, as the logic used is not a single value:

```c++
int crewCount = ship.Crew();
if(!isFlagship)
	crewCount = min(crewCount, ship.RequiredCrew());
string crew = (ship.IsParked() ? "Parked" : to_string(crewCount));
```

## Screenshots
Before: no tooltip.
After:
Screenshot 2024-06-23 at 13 24 24

## Testing Done
I checked that it works as expected.

## Save File
Any save will do.

## Wiki Update
N/A

## Performance Impact
N/A

### CODE REVIEW

File: data/_ui/tooltips.txt
Comment (TheGiraffe3) [inline data/_ui/tooltips.txt:1471]:
```suggestion
	`Crew currently aboard the ship. This number may be fewer than minimum complement.`
```
What about this?

Comment (nickshanks) [reply data/_ui/tooltips.txt:1471]: I would have either
> May be fewer than minimum complement.
or
> This number may be fewer than _the_ minimum complement.
but my preference would be for brevity. this is a treansient tooltip not prose.

Comment (TheGiraffe3) [reply data/_ui/tooltips.txt:1471]: Forgot the the, correct.

File: source/PlayerInfoPanel.h
Comment (warp-core) [inline source/PlayerInfoPanel.h:22]: Is this used by this file?

File: source/PlayerInfoPanel.h
Comment (warp-core) [inline source/PlayerInfoPanel.h:97]: Storing a copy of the SortableColumn object here doesn't seem like a good idea.
If you need access to the entire object, store a (const) pointer to it.

File: source/PlayerInfoPanel.h
Comment (warp-core) [inline source/PlayerInfoPanel.h:65]: This should probably be in the anonymous namespace in PlayerInfoPanel.cpp; I don't it being called from anywhere that doesn't have access to that scope.

File: source/PlayerInfoPanel.cpp
Comment (warp-core) [inline source/PlayerInfoPanel.cpp:822]:
```suggestion
	if(text.empty())
```

File: source/PlayerInfoPanel.cpp
Comment (warp-core) [inline source/PlayerInfoPanel.cpp:962]: This should be returning a const reference.
```suggestion
const string &PlayerInfoPanel::SortableColumn::Tooltip() const
```

File: source/PlayerInfoPanel.h
Comment (warp-core) [inline source/PlayerInfoPanel.h:80]:
```suggestion
		const std::string &Tooltip() const;
```

File: source/ShopPanel.cpp
Comment (nickshanks) [reply source/ShopPanel.cpp:71]: constexpr is for things (usually functions or objects) that can be evaluated at compile time. this does not need to be evaluated, a simple const will suffice (and it fits with all the other code, that just uses const for ints)

File: source/PlayerInfoPanel.h
Comment (nickshanks) [reply source/PlayerInfoPanel.h:22]: I get red squigglies without it. The tooltip needs a foreground and background colour. I only use `Color` as a variable. I could inline it or perhaps declare that a class `Color` exists as is done in some header files. Which would be your preference?

File: source/PlayerInfoPanel.h
Comment (nickshanks) [reply source/PlayerInfoPanel.h:97]: Ahh good point.

File: source/PlayerInfoPanel.h
Comment (nickshanks) [reply source/PlayerInfoPanel.h:65]: OK.

File: source/PlayerInfoPanel.h
Comment (nickshanks) [reply source/PlayerInfoPanel.h:22]: I I've just realised this was the header file you were looking at. I thought it was the cpp file; the function did originally have five arguments, with fg/bg colour being two of them — I copied it from elsewhere, but since it only gets called from one place, there's no need to have so many args.

File: source/ShopPanel.cpp
Comment (TomGoodIdea) [reply source/ShopPanel.cpp:71]: This is unrelated to the new feature and should rather be handled when implementing #9864.

File: source/PlayerInfoPanel.cpp
Comment (TomGoodIdea) [inline source/PlayerInfoPanel.cpp:50]: Why this addition? There are some `Color &` items in this file already.
```suggestion
```

File: source/PlayerInfoPanel.cpp
Comment (TomGoodIdea) [inline source/PlayerInfoPanel.cpp:116]:
```suggestion
```

File: source/PlayerInfoPanel.h
Comment (TomGoodIdea) [inline source/PlayerInfoPanel.h:94]:
```suggestion
	std::vector<ClickZone<const SortableColumn *>> menuZones;
```

File: source/ShopPanel.cpp
Comment (nickshanks) [reply source/ShopPanel.cpp:71]: It was related because that was the code I had copied into the other class. I was on the verge of extracting it into a new Tootip class myself, and made the change because I wanted the two blocks of code to be identical so that they could be identified as the same later on. This change is immaterial to the operation of the program, and should not block the PR IMO.

File: source/PlayerInfoPanel.cpp
Comment (nickshanks) [reply source/PlayerInfoPanel.cpp]: Not sure. I might have had this in a different file at first where it was needed, then decided to move it here.

**Review by TomGoodIdea — APPROVED** [review]
Normally I'd say that the tooltips should trigger after a certain hover time, because that's how all tooltips except for the shop warnings work, but I guess we can leave it for the proposed Tooltip class to fix later.

File: data/_ui/tooltips.txt
Comment (warp-core) [inline data/_ui/tooltips.txt:1471]:
```suggestion
	`Crew currently aboard the ship. May be fewer than required crew.`
```
Or perhaps:
```suggestion
	`Crew currently aboard the ship. May be fewer than required.`
```

**Comment by warp-core** [comment]
> @TomGoodIdea said [this was a pre-requisite](https://github.com/endless-sky/endless-sky/pull/10250#issuecomment-2182656997) for PR #10250 It adds a tooltip to the column headers on the fleet table of the player info panel. I added just one tooltip, for what I think is the least intuitive column currently, crew, as the logic used is not a single value:
> 
> ```c++
> int crewCount = ship.Crew();
> if(!isFlagship)
> 	crewCount = min(crewCount, ship.RequiredCrew());
> string crew = (ship.IsParked() ? "Parked" : to_string(crewCount));
> ```

Is the idea that this column is the least intuitive because sometimes it'll have a number there, and others it'll say "Parked"?
If that is the case, it seems a bit strange that the tooltip doesn't mention the latter, only the former.
I'm not sure the tooltip text as it is adds much. Is it unclear that this is the number of crew aboard the ship?

**Comment by tibetiroka** [comment]
Closing as orphaned. Feel free to reopen this PR once the above comments are addressed.

**Comment by nickshanks** [comment]
Why is this marked "orphaned"?
There's nothing left for me to do. It just needs to be merged.

File: data/_ui/tooltips.txt
Comment (nickshanks) [reply data/_ui/tooltips.txt:1478]: Those are less precise, more ambiguous, and objectivly worse.

**Comment by nickshanks** [comment]
> Is the idea that this column is the least intuitive

That's subjective. I wouldn't say it is or isn't the least intuitive (all are equally intuitive to me).

> because sometimes it'll have a number there, and others it'll say "Parked"?

No. It is to explain that this is the number currently aboard, which is not necearilly the number needed for the ship to function without impairment.

**Comment by nickshanks** [comment]
> because there are unresolved review comments from warp-core

Those are all resolved.

> and you've been inactive for quite a long time

I've been very active being a dad — far too active to work on unpaid stuff.

**Comment by nickshanks** [comment]
> Feel free to reopen this PR

You do know, right, that non-admins on this project do not have permission to re-open their own PRs once they have been closed? I find your presumptive comment very infuriating.

File: source/ShopPanel.cpp
Comment (nickshanks) [reply source/ShopPanel.cpp:71]: These lines got moved to an interface file, so this whole discussion is now moot.

**Comment by TomGoodIdea** [comment]
> > and you've been inactive for quite a long time
> 
> I've been very active being a dad — far too active to work on unpaid stuff.

It's alright, we all work on this unpaid stuff.
But our usual procedure is to close a PR when the OP isn't actively working on the project and the PR requires active maintenance, so that others may have a chance to pick it up and finish in case the OP doesn't return.

> > Feel free to reopen this PR
> 
> You do know, right, that non-admins on this project do not have permission to re-open their own PRs once they have been closed? I find your presumptive comment very infuriating.

We only discovered that relatively recently, so I wouldn't assume any malice in tibetiroka's message.

**Comment by TomGoodIdea** [comment]
Hmm, it seems that even I with Maintainer access to the repo am not able to reopen this PR, so probably Github UI is broken once again.

**Comment by TheGiraffe3** [comment]
~~Unless @Amazinite can reopen it, it's probably best to open a new one.~~
Looks like it can't be reopened, so you will indeed need to create a new PR.

**Pull request closed without merging**
